### PR TITLE
feat: improve how errors are returned

### PR DESCRIPTION
**Pull request by kyranet**
I heard you like readable errors, well, let's take that on a whole different level!

Pending to implement all code and fix tests

### CODE REVIEW

File: src/constraints/BigIntConstraints.ts [hunk 1]
```diff
@@ -1,56 +1,44 @@
...
 import type { IConstraint } from './base/IConstraint';
 import { Comparator, eq, ge, gt, le, lt, ne } from './util/operators';
 
-function bigintComparator(
-	comparator: Comparator,
-	name: string,
-	messageBuilder: ConstraintErrorMessageBuilder<bigint>,
-	number: bigint
-): IConstraint<bigint> {
+function bigintComparator(comparator: Comparator, name: string, expected: string, number: bigint): IConstraint<bigint> {
```
Comment (favna) [inline src/constraints/BigIntConstraints.ts:6]: Is it possible to make these comparator `name` parameters strictly typed?

I'd prefer if you retroactively added a type along the lines of `BigIntComparatorNames = 's.bigint.lt' | 's.bigint.le'` etc.

That way this code is less prone to breaking when we revisit it 6 months from now. Same goes for all of these such array and all that. Alternatively if you have another way to type it strictly that's fine too, just as long as where we call the comparators we have strict typing and autocomplete. For example an alternative might be having one shared interface for all comparators and accessing the string union for this specific one with `SharedInterface['bigint']`

**Review by favna — CHANGES_REQUESTED** [review]

File: src/constraints/BigIntConstraints.ts (hunk 1 above)
Comment (kyranet) [reply src/constraints/BigIntConstraints.ts]: Addressed that in 41bebea.

File: src/constraints/DateConstraints.ts [hunk 2]
```diff
@@ -3,7 +3,10 @@ import { Result } from '../lib/Result';
 import type { IConstraint } from './base/IConstraint';
 import { Comparator, eq, ge, gt, le, lt, ne } from './util/operators';
 
-function dateComparator(comparator: Comparator, name: string, expected: string, number: number): IConstraint<Date> {
+export type DateComparatorConstraintName = `s.date.${'lt' | 'le' | 'gt' | 'ge' | 'eq' | 'ne'}`;
+export type DateConstraintName = DateComparatorConstraintName | `s.date.${'eq(NaN)' | 'ne(NaN)'}`;
```
Comment (favna) [inline src/constraints/DateConstraints.ts:7]: Kyra are you okay? Same goes for all the other ones where you have 2 types.

19-49-25_03-02-2022_vivaldi

19-49-30_03-02-2022_vivaldi

19-50-12_03-02-2022_ApplicationFrameHost

**Review by favna — CHANGES_REQUESTED** [review]

**Review by favna — DISMISSED** [review]

File: src/constraints/BooleanConstraints.ts [hunk 3]
```diff
@@ -2,18 +2,20 @@ import { ConstraintError } from '../lib/errors/ConstraintError';
...
 import type { IConstraint } from './base/IConstraint';
 
+export type BooleanConstraintName = `s.boolean.${boolean}`;
+
 export const booleanTrue: IConstraint<boolean, true> = {
 	run(input: boolean) {
 		return input //
 			? Result.ok(input)
-			: Result.err(new ConstraintError('booleanTrue', 'Expected boolean to be true, but received false', input, true));
+			: Result.err(new ConstraintError('s.boolean.true', 'Invalid boolean value', input, 'expected'));
```
Comment (vladfrangu) [inline src/constraints/BooleanConstraints.ts:11]: How do these errors look for bool?

File: tests/lib/errors/ConstraintError.test.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	});
+
+	describe('inspect', () => {
+		test('GIVEN an inspected instance THEN formats data correctly', () => {
+			const content = inspect(error, { colors: false });
+			const expected = [
+				'ConstraintError > s.number.int', //
+				'  Given value is not an integer',
+				'',
+				'  Expected Number.isInteger(expected)',
```
Comment (vladfrangu) [inline tests/lib/errors/ConstraintError.test.ts:21]: This is an useless error message. `Expected value to be an integer` instead (or `Expected Number.isInteger(value) to be true`

File: tests/lib/errors/ExpectedValidationError.test.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+				''
+			];
+
+			expect(content.startsWith(expected.join('\n'))).toBe(true);
+		});
+
+		test('GIVEN an inspected instance with negative depth THEN formats name only', () => {
+			const content = inspect(error, { colors: false, depth: -1 });
+			const expected = [
+				'[ExpectedValidationError]' //
```
Comment (vladfrangu) [inline tests/lib/errors/ExpectedValidationError.test.ts:35]: at least always include the subsequent error name (`[ExpectedValidationError > LiteralValidator]` or just `[LiteralValidator]`

File: tests/lib/errors/ValidationError.test.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+				''
+			];
+
+			expect(content.startsWith(expected.join('\n'))).toBe(true);
+		});
+
+		test('GIVEN an inspected instance with negative depth THEN formats name only', () => {
+			const content = inspect(error, { colors: false, depth: -1 });
+			const expected = [
+				'[ValidationError]' //
```
Comment (vladfrangu) [inline tests/lib/errors/ValidationError.test.ts:31]: Ditto, specify the validator too

File: tests/validators/number.test.ts [hunk 7]
```diff
@@ -19,47 +19,102 @@ describe('NumberValidator', () => {
...
+				expect(eqPredicate.parse(value)).toBe(value);
+			});
+
+			test.each([10, 100])('GIVEN %d THEN throws ConstraintError', (value) => {
+				expect(() => eqPredicate.parse(value)).toThrow(new ConstraintError('s.number.eq', 'Invalid number value', value, 'expected === 42'));
+			});
+		});
+
+		describe('eq(NaN)', () => {
+			const eqNanPredicate = s.number.eq(NaN);
```
Comment (vladfrangu) [inline tests/validators/number.test.ts:80]: Impressed this works

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/constraints/BooleanConstraints.ts (hunk 3 above)
Comment (kyranet) [reply src/constraints/BooleanConstraints.ts]: image

File: tests/lib/errors/ConstraintError.test.ts (hunk 4 above)
Comment (kyranet) [reply tests/lib/errors/ConstraintError.test.ts]: image

**Review by vladfrangu — APPROVED** [review]
Nice

**Review by favna — APPROVED** [review]

**Pull request merged**